packages/react-keytips/src/hooks/useTree.ts

```diff
@@ -0,0 +1,168 @@
+import type { KeytipProps, KeytipWithId } from '../components/Keytip';
+import * as React from 'react';
+import { sequencesToID, createNode } from '../utilities';
+import { KTP_ROOT_ID } from '../constants';
+import { useFluent } from '@fluentui/react-components';
+
+export type KeytipTreeNode = Pick<
+  KeytipProps,
+  'keySequences' | 'onExecute' | 'onReturn' | 'dynamic'
+> & {
+  id: string;
+  uniqueId: string;
+  target: HTMLElement | null;
+  parent: string;
+  children: Set<string>;
+};
+
+const getParentID = (sequence: string[]): string =>
+  sequence.length === 0
+    ? KTP_ROOT_ID
+    : sequencesToID(sequence.slice(0, sequence.length - 1));
+
+export function useTree() {
+  const { targetDocument } = useFluent();
+  const root: KeytipTreeNode = React.useMemo(
+    () => ({
+      id: KTP_ROOT_ID,
+      uniqueId: KTP_ROOT_ID,
+      children: new Set(),
+      target: null,
+      parent: '',
+      keySequences: [],
+    }),
+    []
+  );
+
+  const nodeMap = React.useRef<Map<string, KeytipTreeNode>>(
+    new Map([[KTP_ROOT_ID, root]])
+  );
+
+  const currentKeytip = React.useRef<KeytipTreeNode | undefined>();
+
+  const addNode = React.useCallback((keytip: KeytipWithId) => {
```

Comment:
@dmytrokirpa Thanks for suggestion. Tbh, not sure, if it will make it easier to test, it will be similar to use useState so I will need to wrap all these methods in tests with `act`, the current implementation with `useRef` also avoids re-renders and not sure if having state here is really needed 🙂

Reply:
yep, but I dont want it to be stateful :)